2.9.0/final.md

```diff
@@ -0,0 +1,832 @@
+The expensive `torch.export.draft_export` diagnostic path is no longer auto-invoked (which could take hours on large models). You can still opt in for deep diagnostics:
+
+Previously in torch 2.8.0:
+
+```bash
+# If both torch.export.export(..., strict=False) and
+# torch.export.export(..., strict=True) fail to capture
+# the model graph, torch.export.draft_export(...) will be triggered,
+# and uses real tensor to trace/export the model.
+#
+# Inside export_to_onnx.py:
+#  ... torch.onnx.export(..., dynamo=True)
+python export_to_onnx.py
+```
+
+Now in torch 2.9.0:
+
+```bash
+# To trigger torch.export.draft_export once
+# torch.export.export strict=False/True both
+# fail:
+
+TORCH_ONNX_ENABLE_DRAFT_EXPORT=True python export_to_onnx.py
+```
+
+### Remove `torch.onnx.dynamo_export` and the `onnxrt` torch compile backend ([#158130](https://github.com/pytorch/pytorch/pull/158130), [#158258](https://github.com/pytorch/pytorch/pull/158258))
+
+`torch.onnx.dynamo_export` is removed. Please use `torch.onnx.export` instead.
+The experimental ONNX Runtime compile backend (`torch.compile(backend="onnxrt")`) is no longer supported.
+
+### Remove `torch.onnx.enable_fake_mode` ([#161222](https://github.com/pytorch/pytorch/pull/161222))
+
+The `dynamo=True` mode uses `FakeTensor`s by default which is memory efficient.
+
+### Some public facing utility APIs for the TorchScript based exporter are now private ([#161323](https://github.com/pytorch/pytorch/pull/161323))
+### Remove `torch.onnx.symbolic_caffe2` ([#157102](https://github.com/pytorch/pytorch/pull/157102))
+
+## Python Frontend
+### Upgrade to DLPack 1.0. ([#145000](https://github.com/pytorch/pytorch/pull/145000))
+
+This upgrade is doing the same BC-breaking changes as the DLPack release.
+Objects in `torch.utils.dlpack` have been updated to reflect these changes, such as `DLDeviceType`.
+See the PR for details on the exact changes and how to update your code.
+
+### Raise appropriate errors in `torch.cat` ([#158249](https://github.com/pytorch/pytorch/pull/158249))
+
+Raising `ValueError`, `IndexError` or `TypeError` where appropriate instead of the generic `RuntimeError`.
+If you code was catching these error, you can update to catch the new error type.
+
+# Deprecations
+## Dataloader Frontend
+### Deprecate `pin_memory_device` param in `torch.utils.data.DataLoader` ([#158323](https://github.com/pytorch/pytorch/pull/158323))
+
+We move enabling `pin_memory` back inside `BaseDataLoaderIter`. This is required for `StatefulDataloader` which leveraged `BaseDataLoaderIter` direclty rather than the `Dataloader` class init
+
+## Export
+### Deprecation for `export_for_training` API, in favor of equivalent `export` API ([#158203](https://github.com/pytorch/pytorch/pull/158203))
+
+`export_for_training` exists because we couldn't migrate internal usages of export to the final IR. Now that we have completed the migration, we deprecated and deleted this API.
```

Comment:
Mention the long API (like not just export, but torch.blah.blah.export) that replaces this API (also include the long version of this API)